### PR TITLE
Mmakoviichuk iftaqcaa 3 - Categories test

### PR DESCRIPTION
**Pull request by Ferrl-m**

### CODE REVIEW

File: src/main/java/com/softserveinc/ita/pageobjects/CategoriesPage.java [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+    private final String subCategorySelector = "(//a[contains(@class , 'tile-cats__heading')])[1]";
+    private final String productSelector = "(//div[@class = 'goods-tile__inner'])[1]";
+
+    public void openSubCategory() {
+        $x(subCategorySelector).click();
+    }
+
+    public void openProduct() {
+        $x(productSelector).click();
+    }
```
Comment (PKuravskyi) [inline src/main/java/com/softserveinc/ita/pageobjects/CategoriesPage.java:16]: maybe move this to ProductsPage?

File: src/main/java/com/softserveinc/ita/pageobjects/CategoriesPage.java [hunk 2]
```diff
@@ -0,0 +1,17 @@
+package com.softserveinc.ita.pageobjects;
+
+import static com.codeborne.selenide.Selenide.$x;
+
+public class CategoriesPage extends BasePage<CategoriesPage>{
+
+    private final String subCategorySelector = "(//a[contains(@class , 'tile-cats__heading')])[1]";
```
Comment (Nikodych) [inline src/main/java/com/softserveinc/ita/pageobjects/CategoriesPage.java:7]: need to review xpath's
I will do it later and suggest something

File: src/main/java/com/softserveinc/ita/pageobjects/ProductPage.java [hunk 3]
```diff
@@ -0,0 +1,12 @@
+package com.softserveinc.ita.pageobjects;
+
+import static com.codeborne.selenide.Selenide.$x;
+
+public class ProductPage extends BasePage<ProductPage>{
+
+    private final String categorySelector = "//ul[@class = 'breadcrumbs ng-star-inserted']";
```
Comment (Nikodych) [inline src/main/java/com/softserveinc/ita/pageobjects/ProductPage.java:7]: productCategorySelector

File: src/main/java/com/softserveinc/ita/utils/runners/TestRunner.java [hunk 4]
```diff
@@ -1,16 +1,20 @@
...
+import com.softserveinc.ita.pageobjects.CategoriesPage;
 import com.softserveinc.ita.pageobjects.HomePage;
+import com.softserveinc.ita.pageobjects.ProductPage;
 import org.testng.annotations.BeforeMethod;
 
 import static com.codeborne.selenide.Selenide.open;
 
 public abstract class TestRunner {
 
+    protected CategoriesPage categoriesPage = new CategoriesPage();
```
Comment (Nikodych) [inline src/main/java/com/softserveinc/ita/utils/runners/TestRunner.java:12]: init pages only in tests or get them through return statements in methods

delete pages from here, you can init homepage in homepagetest class

File: src/test/java/com/softserveinc/ita/HomePageTest.java [hunk 5]
```diff
@@ -9,6 +11,34 @@
...
+
+    @Test(dataProvider = "rozetkaCategoryData")
+    public void dropdownCategoryTest(String title) {
+        CatalogModal catalogModal = categoriesPage.openCatalog();
+        catalogModal.openDropdownCategory(title);
+        categoriesPage.openSubCategory();
+        categoriesPage.openProduct();
+
+        assertThat(productPage.isCategoryCorrect(title))
+                .as("Product should correspond " + title + " category");
```
Comment (Nikodych) [inline src/test/java/com/softserveinc/ita/HomePageTest.java:40]: assert is not completed

File: src/test/java/com/softserveinc/ita/HomePageTest.java [hunk 6]
```diff
@@ -9,6 +11,34 @@
...
+    }
     @Test
```
Comment (Nikodych) [inline src/test/java/com/softserveinc/ita/HomePageTest.java:42]: add empty line

**Review by Nikodych — CHANGES_REQUESTED** [review]

File: src/main/java/com/softserveinc/ita/pageobjects/CategoriesPage.java (hunk 2 above)
Comment (Nikodych) [reply src/main/java/com/softserveinc/ita/pageobjects/CategoriesPage.java:7]: if I understand, it always will be the first subcategory? probably better to specify or through random chose category

File: src/main/java/com/softserveinc/ita/pageobjects/CategoriesPage.java [hunk 7]
```diff
@@ -7,11 +7,15 @@
     private final String subCategorySelector = "(//a[contains(@class , 'tile-cats__heading')])[1]";
     private final String productSelector = "(//div[@class = 'goods-tile__inner'])[1]";
```
Comment (Nikodych) [inline src/main/java/com/softserveinc/ita/pageobjects/CategoriesPage.java:8]: same

File: src/test/java/com/softserveinc/ita/HomePageTest.java [hunk 8]
```diff
@@ -20,26 +23,28 @@
...
     public void categoryTest(String title) {
-        homePage.openCategory(title);
-        categoriesPage.openSubCategory();
-        categoriesPage.openProduct();
+         homePage.openCategory(title)
+                .openSubCategory()
+                .openProduct();
 
-        assertThat(productPage.isCategoryCorrect(title))
+        assertThat(new ProductPage().getProductCategory().contains(title))
```
Comment (Nikodych) [inline src/test/java/com/softserveinc/ita/HomePageTest.java:30]: extract into variable. probably get category name and compare with expected

File: src/test/java/com/softserveinc/ita/HomePageTest.java [hunk 9]
```diff
@@ -20,26 +23,28 @@
     }
     @Test(dataProvider = "rozetkaCategoryData")
     public void categoryTest(String title) {
-        homePage.openCategory(title);
-        categoriesPage.openSubCategory();
-        categoriesPage.openProduct();
+         homePage.openCategory(title)
```
Comment (Nikodych) [inline src/test/java/com/softserveinc/ita/HomePageTest.java:26]: first method should be on next line

homePage
.openCategory(title)

File: src/test/java/com/softserveinc/ita/HomePageTest.java [hunk 10]
```diff
@@ -20,26 +23,28 @@
...
                 .isTrue();
     }
 
     @Test(dataProvider = "rozetkaCategoryData")
     public void dropdownCategoryTest(String title) {
-        CatalogModal catalogModal = categoriesPage.openCatalog();
-        catalogModal.openDropdownCategory(title);
-        categoriesPage.openSubCategory();
-        categoriesPage.openProduct();
+        homePage.openCatalog()
```
Comment (Nikodych) [inline src/test/java/com/softserveinc/ita/HomePageTest.java:37]: same

File: src/test/java/com/softserveinc/ita/HomePageTest.java [hunk 11]
```diff
@@ -20,26 +23,28 @@
     }
     @Test(dataProvider = "rozetkaCategoryData")
     public void categoryTest(String title) {
```
Comment (Nikodych) [inline src/test/java/com/softserveinc/ita/HomePageTest.java:25]: verifyCategoryTransitionTest

File: src/test/java/com/softserveinc/ita/HomePageTest.java [hunk 12]
```diff
@@ -20,26 +23,28 @@
...
-        categoriesPage.openSubCategory();
-        categoriesPage.openProduct();
+        homePage.openCatalog()
+                .openDropdownCategory(title)
+                .openSubCategory()
+                .openProduct();
 
-        assertThat(productPage.isCategoryCorrect(title))
-                .as("Product should correspond " + title + " category");
+        assertThat(new ProductPage().getProductCategory().contains(title))
```
Comment (Nikodych) [inline src/test/java/com/softserveinc/ita/HomePageTest.java:42]: same

File: src/test/java/com/softserveinc/ita/HomePageTest.java [hunk 13]
```diff
@@ -20,26 +23,28 @@
...
+                .openProduct();
 
-        assertThat(productPage.isCategoryCorrect(title))
+        assertThat(new ProductPage().getProductCategory().contains(title))
                 .as("Product should correspond " + title + " category")
                 .isTrue();
     }
 
     @Test(dataProvider = "rozetkaCategoryData")
     public void dropdownCategoryTest(String title) {
```
Comment (Nikodych) [inline src/test/java/com/softserveinc/ita/HomePageTest.java:36]: verifyCategoryTransitionThroughDropdownTest

**Review by Nikodych — CHANGES_REQUESTED** [review]

File: src/main/java/com/softserveinc/ita/pageobjects/CategoriesPage.java (hunk 7 above)
Comment (Ferrl-m) [reply src/main/java/com/softserveinc/ita/pageobjects/CategoriesPage.java:8]: But testcase says that we need first category
image

File: src/test/java/com/softserveinc/ita/HomePageTest.java [hunk 14]
```diff
@@ -22,29 +22,35 @@
...
+        homePage
+                .openCatalog()
                 .openDropdownCategory(title)
                 .openSubCategory()
                 .openProduct();
 
-        assertThat(new ProductPage().getProductCategory().contains(title))
+        var actualTitle = new ProductPage().getProductCategory();
+
+        assertThat(actualTitle.contains(title))
```
Comment (Nikodych) [inline src/test/java/com/softserveinc/ita/HomePageTest.java:48]: assertThat(actualTitle)
.as("")
.isEqual(title);

File: src/test/java/com/softserveinc/ita/HomePageTest.java [hunk 15]
```diff
@@ -22,29 +22,35 @@
...
+    public void verifyCategoryTransitionTest(String title) {
+         homePage
+                 .openCategory(title)
                 .openSubCategory()
                 .openProduct();
 
-        assertThat(new ProductPage().getProductCategory().contains(title))
+         var actualTitle = new ProductPage().getProductCategory();
+
+        assertThat(actualTitle.contains(title))
```
Comment (Nikodych) [inline src/test/java/com/softserveinc/ita/HomePageTest.java:33]: assertThat(actualTitle)
.as("")
.isEqual(title);

File: src/test/java/com/softserveinc/ita/HomePageTest.java [hunk 16]
```diff
@@ -22,29 +22,35 @@
         };
     }
     @Test(dataProvider = "rozetkaCategoryData")
-    public void categoryTest(String title) {
-         homePage.openCategory(title)
+    public void verifyCategoryTransitionTest(String title) {
+         homePage
+                 .openCategory(title)
```
Comment (Nikodych) [inline src/test/java/com/softserveinc/ita/HomePageTest.java:27]: format code

**Review by Nikodych — CHANGES_REQUESTED** [review]

**Review by Nikodych — APPROVED** [review]

**Pull request merged**